sklearn/feature_selection/_correlation_threshold.py

```diff
@@ -0,0 +1,214 @@
+    >>> X = np.array([[0.0, 1.0, 2.0], [1.1, 2.0, 3.0], [0.5, 10.1, 1.1]]).T
+    >>> selector = CorrelationThreshold(threshold=0.9)
+    >>> selector.fit_transform(X)
+    array([[ 1.1,  0.5],
+           [ 2. , 10.1],
+           [ 3. ,  1.1]])
+
+    See Also
+    --------
+    :class:`sklearn.cluster.FeatureAgglomeration` : Agglomerate features by
+        recursively merges features
+
+    References
+    ----------
+    Max Kuhn and Kjell Johnson, "Applied Predictive Modeling", Springer, 2013
+    (Section 3.5)
+    """
+    def __init__(self, threshold=0.9, kind='pearson'):
+        self.threshold = threshold
+        self.kind = kind
+
+    def fit(self, X, y=None):
+        """Learn empirical variances from X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training set to compute correlations.
+
+        y : ignored
+            Not used, present here for API consistency by convention.
+
+        Returns
+        -------
+        self : `CorrelationThreshold`
+        """
+        if not (0.0 <= self.threshold <= 1.0):
+            raise ValueError("threshold must be in [0.0, 1.0], got {}".format(
+                             self.threshold))
+        if self.kind not in ('pearson', 'spearmanr'):
+            raise ValueError("kind must be 'pearson' or 'spearmanr'")
+
+        if issparse(X) and self.kind != 'pearson':
+            raise ValueError("only pearson correlation is supported with "
+                             "sparse matrices")
+
+        X = check_array(X, accept_sparse=['csc', 'csr'], dtype=[np.float64,
+                                                                np.float32])
+
+        n_features = X.shape[1]
+        if self.threshold == 1.0 or (1 in X.shape):
+            self.support_mask_ = np.ones(n_features, dtype=np.bool)
+            return self
+
+        # get constant features
+        if issparse(X):
+            mins, maxes = min_max_axis(X, axis=0)
+            peak_to_peaks = maxes - mins
+            constant_mask = np.isclose(peak_to_peaks, 0.0)
+
+            # sparse correlation
+            mu, sparse_var = mean_variance_axis(X, 0)
+            X_corr = _sparse_correlation(X, mu, ~constant_mask)
+        else:
+            peak_to_peaks = np.ptp(X, axis=0)
+            constant_mask = np.isclose(peak_to_peaks, 0.0)
+
+            if self.kind == 'pearson':
+                X_corr = np.corrcoef(X, rowvar=False)
+            else:  # spearmanr
+                X_corr, _ = spearmanr(X)
+                # spearmanr returns scaler when comparing two columns
+                if isinstance(X_corr, float):
+                    X_corr = np.array([[1, X_corr], [X_corr, 1]])
+
+        np.fabs(X_corr, out=X_corr)
+
+        # Removes constant features from support_mask
+        support_mask = np.ones(n_features, dtype=np.bool)
+        upper_idx = np.triu_indices(n_features, 1)
+
```

Comment:
Should we not immediately remove from upper_idx coordinates where `X_corr < threshold`?